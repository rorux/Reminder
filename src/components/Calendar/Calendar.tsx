import React from 'react';

const Calendar = () => {
  return <div data-test="calendar">Календарь</div>;
};

export default Calendar;
