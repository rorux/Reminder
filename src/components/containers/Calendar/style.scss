.months-wrap {
  display: flex;
  justify-content: center;
}
