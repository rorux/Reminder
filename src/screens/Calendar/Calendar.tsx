import React from 'react';

const Calendar = () => {
  return (
    <div className="calendar" data-test="calendarComponent">
      Календарь
    </div>
  );
};

export default Calendar;
