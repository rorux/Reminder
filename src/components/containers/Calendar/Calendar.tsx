import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import Paper from '@mui/material/Paper';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import { grey } from '@mui/material/colors';
import { NavigateBefore, NavigateNext } from '@mui/icons-material';
import { CircularProgress } from '@mui/material';
import { initRecords } from '@store/records/actions';
import Month from '../Month';
import { getMonthName } from './constants';
import { recordsSelector } from '@store/records/selectors';
import './style.scss';
import Modal from '@components/Modal';

const todayDate = new Date();
const todayParse = Date.parse(
  String(new Date(todayDate.getFullYear(), todayDate.getMonth(), todayDate.getDate()))
);
export const MyContext = React.createContext({ todayParse });

const Calendar = () => {
  const [year, setYear] = useState(new Date().getFullYear());
  const [month, setMonth] = useState(new Date().getMonth());
  const [isBeforeMonthDisabled, setIsBeforeMonthDisabled] = useState(false);
  const [isNextMonthDisabled, setIsNextMonthDisabled] = useState(false);

  const { recordList, loading } = useSelector(recordsSelector);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(initRecords());
  }, [dispatch]);

  useEffect(() => {
    if (month === 0) setIsBeforeMonthDisabled(true);
    else setIsBeforeMonthDisabled(false);
    if (month === 11) setIsNextMonthDisabled(true);
    else setIsNextMonthDisabled(false);
  }, [month]);

  const handleBeforeMonth = () => {
    setMonth((month) => month - 1);
  };

  const handleNextMonth = () => {
    setMonth((month) => month + 1);
  };

  return (
    <>
      {loading ? (
        <CircularProgress />
      ) : (
        <div data-test="calendar" className="months-wrap">
          <Box
            sx={{
              minWidth: 260,
              backgroundColor: grey[200],
              borderRadius: 1,
              boxSizing: 'border-box',
            }}
            p={1}
          >
            <Paper elevation={0} sx={{ padding: '5px', paddingTop: '2px' }}>
              <Grid sx={{ height: 40 }} p={1} container>
                <Grid item sm flexGrow={1}>
                  <Typography variant="overline" gutterBottom>
                    {getMonthName[month]} {year}
                  </Typography>
                </Grid>
                <Grid item>
                  <IconButton
                    disabled={isBeforeMonthDisabled}
                    aria-label="delete"
                    size="small"
                    onClick={handleBeforeMonth}
                  >
                    <NavigateBefore fontSize="inherit" />
                  </IconButton>
                  <IconButton
                    disabled={isNextMonthDisabled}
                    aria-label="delete"
                    size="small"
                    onClick={handleNextMonth}
                  >
                    <NavigateNext fontSize="inherit" />
                  </IconButton>
                </Grid>
              </Grid>
              <MyContext.Provider value={{ todayParse }}>
                <Month year={year} month={month} recordList={recordList} />
              </MyContext.Provider>
            </Paper>
          </Box>
          <Modal />
        </div>
      )}
    </>
  );
};

export default Calendar;
