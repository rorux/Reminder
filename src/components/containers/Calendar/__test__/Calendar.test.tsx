import React from 'react';
import * as Redux from 'react-redux';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
// import 'jsdom-global/register';
import Enzyme, { shallow } from 'enzyme';
import { CircularProgress } from '@mui/material';
import Calendar from '../Calendar';

Enzyme.configure({ adapter: new Adapter() });

const recordListParam = [
  {
    id: '1',
    title: 'title of monthly remind',
    period: 'monthly',
    monthday: '8',
    holidays: 'afterHoliday',
    days: [1642366800000, 1644872400000],
  },
  {
    id: '2',
    title: 'title of quarterly remind',
    period: 'quarterly',
    monthday: '8',
    quarter: 'Янв, Апр, Июл, Окт',
    holidays: 'beforeHoliday',
    days: [1642366800000, 1644872400000],
  },
];

const createSelector = (loading: boolean, recordList = recordListParam) => ({
  recordList,
  loading,
  error: false,
});

describe('Dashboard component', () => {
  let useEffect: any;
  let useDispatch: any;
  let useSelector: any;
  let setState: any;
  let useStateSpy;
  let mockDispatch: jest.Mock;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f: any) => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    useDispatch = jest.spyOn(Redux, 'useDispatch');
    useSelector = jest.spyOn(Redux, 'useSelector');
    mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    setState = jest.fn();
    useStateSpy = jest.spyOn(React, 'useState');
    // @ts-ignore
    useStateSpy.mockImplementation((init) => [init, setState]);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('Snapshot of Dashboard component', () => {
    useSelector.mockReturnValue(createSelector(false));
    const component = shallow(<Calendar />);
    expect(component).toMatchSnapshot();
  });

  it('Should work "useEffect"', () => {
    useSelector.mockReturnValue(createSelector(false));
    mockUseEffect();
    shallow(<Calendar />);
    expect(mockDispatch).toHaveBeenCalled();
  });

  it('Should render CircularProgress when loading', () => {
    useSelector.mockReturnValue(createSelector(true));
    const component = shallow(<Calendar />);
    expect(component.find(CircularProgress)).toHaveLength(1);
  });

  it('Should change state when click "button-next"', () => {
    useSelector.mockReturnValue(createSelector(false));
    const component = shallow(<Calendar />);
    // @ts-ignore
    component.find(`[data-test='button-next']`).props().onClick();
    expect(setState).toHaveBeenCalled();
  });

  it('Should change state when click "button-before"', () => {
    useSelector.mockReturnValue(createSelector(false));
    const component = shallow(<Calendar />);
    // @ts-ignore
    component.find(`[data-test='button-before']`).props().onClick();
    expect(setState).toHaveBeenCalled();
  });

  it('Should work correct when month is january', () => {
    // @ts-ignore
    jest.spyOn(global, 'Date').mockImplementation(() => ({
      getFullYear: jest.fn(() => 2022),
      getMonth: jest.fn(() => 0),
    }));

    useSelector.mockReturnValue(createSelector(false));
    mockUseEffect();
    mockUseEffect();

    const component = shallow(<Calendar />);
    // @ts-ignore
    component.find(`[data-test='button-next']`).props().onClick();
    expect(setState).toHaveBeenCalledWith(1);
  });

  it('Should work correct when month is december', () => {
    // @ts-ignore
    jest.spyOn(global, 'Date').mockImplementation(() => ({
      getFullYear: jest.fn(() => 2022),
      getMonth: jest.fn(() => 11),
    }));

    useSelector.mockReturnValue(createSelector(false));
    mockUseEffect();
    mockUseEffect();

    const component = shallow(<Calendar />);
    // @ts-ignore
    component.find(`[data-test='button-before']`).props().onClick();
    expect(setState).toHaveBeenCalledWith(10);
  });
});
