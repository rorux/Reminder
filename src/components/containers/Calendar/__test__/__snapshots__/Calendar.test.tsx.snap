// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component Snapshot of Dashboard component 1`] = `
<Fragment>
  <div
    className="months-wrap"
    data-test="calendar"
  >
    <ForwardRef(Box)
      p={1}
      sx={
        Object {
          "backgroundColor": "#eeeeee",
          "borderRadius": 1,
          "boxSizing": "border-box",
          "minWidth": 260,
        }
      }
    >
      <ForwardRef(Paper)
        elevation={0}
        sx={
          Object {
            "padding": "5px",
            "paddingTop": "2px",
          }
        }
      >
        <ForwardRef(Grid)
          container={true}
          p={1}
          sx={
            Object {
              "height": 40,
            }
          }
        >
          <ForwardRef(Grid)
            flexGrow={1}
            item={true}
            sm={true}
          >
            <ForwardRef(Typography)
              gutterBottom={true}
              variant="overline"
            >
              Январь
               
              2022
            </ForwardRef(Typography)>
          </ForwardRef(Grid)>
          <ForwardRef(Grid)
            item={true}
          >
            <ForwardRef(IconButton)
              aria-label="delete"
              data-test="button-before"
              disabled={false}
              onClick={[Function]}
              size="small"
            >
              <Memo(ForwardRef(NavigateBeforeIcon))
                fontSize="inherit"
              />
            </ForwardRef(IconButton)>
            <ForwardRef(IconButton)
              aria-label="delete"
              data-test="button-next"
              disabled={false}
              onClick={[Function]}
              size="small"
            >
              <Memo(ForwardRef(NavigateNextIcon))
                fontSize="inherit"
              />
            </ForwardRef(IconButton)>
          </ForwardRef(Grid)>
        </ForwardRef(Grid)>
        <ContextProvider
          value={
            Object {
              "todayParse": 1642194000000,
            }
          }
        >
          <Month
            month={0}
            recordList={
              Array [
                Object {
                  "days": Array [
                    1642366800000,
                    1644872400000,
                  ],
                  "holidays": "afterHoliday",
                  "id": "1",
                  "monthday": "8",
                  "period": "monthly",
                  "title": "title of monthly remind",
                },
                Object {
                  "days": Array [
                    1642366800000,
                    1644872400000,
                  ],
                  "holidays": "beforeHoliday",
                  "id": "2",
                  "monthday": "8",
                  "period": "quarterly",
                  "quarter": "Янв, Апр, Июл, Окт",
                  "title": "title of quarterly remind",
                },
              ]
            }
            year={2022}
          />
        </ContextProvider>
      </ForwardRef(Paper)>
    </ForwardRef(Box)>
    <CustomizedDialogs />
  </div>
</Fragment>
`;
